controllers/main.go

```diff
@@ -138,54 +139,47 @@ func NewMainController(params *chaincfg.Params, adminIPs []string, baseURL strin
 }
 
 // API is the main frontend that handles all API requests.
-// XXX This is very hacky.  It can be made more elegant by improving or
-// re-writing the middleware and switching to a more modern framework
-// such as goji2 or something else that utilizes the context package/built-in.
-// This would make it much easier to share code between the web UI and JSON
-// API with less duplication.
-func (controller *MainController) API(c web.C, r *http.Request) (string, int) {
-	// we expect /api/vX.YZ/command
-	URIparts := strings.Split(r.RequestURI, "/")
-	if len(URIparts) != 4 {
-		return "{\"status\": \"error\"," +
-			"\"message\": \"invalid API request\"}\n", http.StatusOK
-	}
-	version := URIparts[2]
-	command := URIparts[3]
-
-	if version != "v0.1" {
-		return "{\"status\": \"error\"," +
-			"\"message\": \"invalid API version\"}\n", http.StatusOK
-	}
+func (controller *MainController) API(c web.C, r *http.Request) *system.APIResponse {
+	command := c.URLParams["command"]
+
+	// poolapi.Response comprises a status, message, and a data struct
+	var response, status string
+	var data interface{}
+
+	var err error
 
 	switch r.Method {
 	case "GET":
 		switch command {
 		case "getPurchaseInfo":
-			data, response, err := controller.APIPurchaseInfo(c, r)
-			return APIResponse(data, response, err), http.StatusOK
+			data, response, err = controller.APIPurchaseInfo(c, r)
 		case "startsession":
-			return APIResponse(nil, "session started", nil), http.StatusOK
+			status, response = "success", "session started"
 		case "stats":
-			data, response, err := controller.APIStats(c, r)
-			return APIResponse(data, response, err), http.StatusOK
+			data, response, err = controller.APIStats(c, r)
+		default:
+			return nil
 		}
 	case "POST":
 		switch command {
 		case "address":
-			_, response, err := controller.APIAddress(c, r)
-			return APIResponse(nil, response, err), http.StatusOK
+			_, response, err = controller.APIAddress(c, r)
 		case "signin":
-			_, response, err := controller.APISignIn(c, r)
-			return APIResponse(nil, response, err), http.StatusOK
+			_, response, err = controller.APISignIn(c, r)
 		case "signup":
-			_, response, err := controller.APISignUp(c, r)
-			return APIResponse(nil, response, err), http.StatusOK
+			_, response, err = controller.APISignUp(c, r)
+		default:
+			return nil
 		}
 	}
 
-	return "{\"status\": \"error\"," +
-		"\"message\": \"invalid API command\"}\n", http.StatusOK
+	if err != nil {
+		status = err.Error()
```

Comment:
it's not a massive issue in this case. the paymetheus implementation checks for success rather than checking for error, and assumes error if it's not success.